### PR TITLE
Added GBMB GBR1 (broken) ksy format

### PR DESCRIPTION
**Pull request by ZetaTwo**
As per: https://github.com/ZetaTwo/gameboy-project/issues/1 I am opening this PR.
However, this KSY does not actually properly parse the file format since I have not found a good way to implement it.
The problem is described in: https://github.com/kaitai-io/kaitai_struct/issues/172

The problem is as follows. The file format consists of a sequence of "Objects". Each such object has (among other things) a Type, an ObjectID and a MasterID. If the MasterID is 0, it is a regular object, otherwise it is a "Sub-object" referencing its parent. In particular, there are two type: Map (type=2) and Map Tile Data (type=3) where the Map has two attributes: width and height and the size of the Map Tile Data object is determined by looking up its master object and repeating a record "width*height" times.

Basically, it would be required to write something like:
```
object_map_tile_data:
    seq:
      - id: master_id
        type: u2
      - id: data
        type: object_map_tile_data_record
        repeat: expr
        #repeat-expr: master.tile_count
        repeat-expr: 0
    instances:
      master:
        value: find_master(master_id)
```

How would you suggest I approach this with Kaitai currently?

### CODE REVIEW

File: game/gbtd_gbr1.ksy [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    seq:
+      - id: master_id
+        type: u2
+      - id: data
+        type: object_map_tile_data_record
+        repeat: expr
+        #repeat-expr: master.tile_count
+        repeat-expr: 0
+    instances:
+      master:
```
Comment (KOLANICH) [inline game/gbtd_gbr1.ksy:98]:
```yaml
-affected-by: 172
```

**Review by KOLANICH — COMMENTED** [review]
The same issues as in #401 are also present here.

**Comment by ZetaTwo** [comment]
Updated PR based on comments in #401

File: game/gbtd_gbr1.ksy [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    contents: GBO1
+  - id: objects
+    type: gbr1_object
+    repeat: eos
+
+types:
+  gbr1_object:
+    seq:
+      - id: marker
+        contents: [0x48, 0x50, 0x4a, 0x4d, 0x54, 0x4c] # HPJMTL
```
Comment (KOLANICH) [inline game/gbtd_gbr1.ksy:29]: `contents: HPJMTL`

Comment (ZetaTwo) [reply game/gbtd_gbr1.ksy]: Fixed

**Comment by ZetaTwo** [comment]
Updated PR based on your comments.

File: game/gbtd_gbr1.ksy [hunk 3]
```diff
@@ -69,103 +69,104 @@ types:
...
+          - id: master_id
+            type: u2
+          - id: data
+            type: map_tile_data_record
+            repeat: expr
+            #repeat-expr: master.tile_count
+            repeat-expr: 0
+        instances:
+          master:
+          #  value: find_master(master_id) # <--- Doesn't work (https://github.com/kaitai-io/kaitai_struct/issues/172)
```
Comment (KOLANICH) [inline game/gbtd_gbr1.ksy:115]: `-affected-by: 172`

**Comment by ZetaTwo** [comment]
Made an implementation using opaque types in the meanwhile. It's not fantastic but it works.

File: game/gbtd_gbr1.ksy [hunk 4]
```diff
@@ -4,6 +4,7 @@ meta:
   encoding: ascii
   title: GBM file
   license: MIT
+  ks-opaque-types: true
```
Comment (KOLANICH) [inline game/gbtd_gbr1.ksy:7]: It'd be nice to insert a link to `lookup_object` source code.

File: game/gbtd_gbr1.ksy [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+              - id: reserved1
+                type: b3
+              - id: sgb_palette
+                type: b3
+              - id: reserved2
+                type: b1
+              - id: gbc_palette
+                type: b5
+              - id: tile_number
+                type: b10
```
Comment (generalmimon) [inline game/gbtd_gbr1.ksy:120]: Does this format really use **little-endian** byte-sized integers and [**big-endian** bit-sized integers](https://doc.kaitai.io/user_guide.html#bit-ints-be)? That would be something quite rare and uncommon. The ["bit endianness"](https://doc.kaitai.io/user_guide.html#bit-endian) is usually the same as byte endianness in the entire format. That's because the bitfields are mostly designed to be packed into a byte-aligned integer, which is first parsed in the same endianness as other byte-aligned integers use, and then the value of such integer is unpacked into the individual bit field members (bit-sized integers) using bitwise AND `&` and SHR `>>` operations.

And reading multi-byte integers in a non-native endianness is mostly less efficient than in the native endianness (since [byte swapping](https://en.wikipedia.org/wiki/Endianness#Byte_swapping) has to be done), so that's something file format authors keep in mind and try to avoid that. Therefore mixing BE and LE integers in the same format does not make sense.

Comment (KOLANICH) [reply game/gbtd_gbr1.ksy:120]: I think it was a big mistake to introduce byte-swapping for it instead of relying on a bit ranges remaper.

File: game/gbtd_gbr1.ksy (hunk 4 above)
Comment (ZetaTwo) [reply game/gbtd_gbr1.ksy:7]: Added link to my implementation in the doc attribute.

File: game/gbtd_gbr1.ksy (hunk 5 above)
Comment (ZetaTwo) [reply game/gbtd_gbr1.ksy:120]: I'm not sure I completely follow here but the current implementation does produce the correct result. Running this test script: https://github.com/ZetaTwo/gameboy-project/blob/master/gbm-parser/tester.py with the test file in that repo gives the correct tile_number values:

```
...

Type: Type.map_tile_data, ID: 2
Size: 429
MapTileData
	Num records: 143
	MapTileData[0].tile_number = 0
	MapTileData[1].tile_number = 1
	MapTileData[2].tile_number = 2
	MapTileData[3].tile_number = 3
	MapTileData[4].tile_number = 4
	...
```

If they were the wrong endianness I would expect it to be something like: "0, 512, 256, 768, ...", right?

**Comment by ZetaTwo** [comment]
Addressed both comments.

**Comment by ZetaTwo** [comment]
Can this be merged?

**Comment by KOLANICH** [comment]
Have you used the wrong order of arguments of `git rebase` command?

**Comment by ZetaTwo** [comment]
I did something wrong indeed but I believe I hade sorted it out now.
